src/main/java/org/commcare/formplayer/postgresutil/PostgresSqlHelper.java

```diff
@@ -50,40 +56,69 @@ public static int setPreparedStatementArgs(PreparedStatement preparedStatement,
      */
-    public static void updateToTable(Connection connection, String tableName, Persistable persistable, int id, String currentSchema) {
-        String queryStart = "UPDATE " + currentSchema + "." + tableName + " SET ";
-        String queryEnd = " WHERE " + DatabaseHelper.ID_COL + " = ?;";
+    public static void upsertToTable(Connection connection, String tableName, Persistable persistable, int id, String currentSchema) {
+        StringBuilder query = new StringBuilder();
+        query.append("INSERT INTO ")
+                .append(currentSchema)
+                .append(".")
+                .append(tableName)
+                .append(" (")
+                .append(DatabaseHelper.ID_COL);
 
         HashMap<String, Object> map = DatabaseHelper.getMetaFieldsAndValues(persistable);
         String[] fieldNames = map.keySet().toArray(new String[map.keySet().size()]);
         Object[] values = map.values().toArray(new Object[map.values().size()]);
+        Object[] params = ArrayUtils.addAll(values, values);
+        for (String field: fieldNames) {
+            query.append(", ")
+                    .append(field);
+        }
 
-        StringBuilder stringBuilder = new StringBuilder(queryStart);
+        query.append(") VALUES (").append(id);
+
+        for (Object obj: values) {
+            query.append(", ?");
+        }
+        query.append(") ON CONFLICT (")
+                .append(DatabaseHelper.ID_COL)
+                .append(") Do UPDATE SET ");
         for (String fieldName : fieldNames) {
-            stringBuilder.append(fieldName).append(" = ?").append(", ");
+            query.append(fieldName).append(" = ?").append(", ");
         }
-        stringBuilder.deleteCharAt(stringBuilder.lastIndexOf(","));
-
-        String query = stringBuilder.append(queryEnd).toString();
+        query.deleteCharAt(query.lastIndexOf(","));
+        query.append(";");
 
-        try (PreparedStatement preparedStatement = connection.prepareStatement(query)){
-            int lastArgIndex = setPreparedStatementArgs(preparedStatement, persistable, values);
-            preparedStatement.setInt(lastArgIndex, id);
+        try (PreparedStatement preparedStatement = connection.prepareStatement(query.toString())){
+            setPreparedStatementArgs(preparedStatement, persistable, params);
             preparedStatement.executeUpdate();
         } catch (SQLException e) {
             throw new SQLiteRuntimeException(e);
         }
     }
 
+
+    /**
+     * Update entry under id with persistable p
+     *
+     * @param connection  Database Connection
+     * @param tableName   name of table
+     * @param persistable persistable to update with
+     * @param id          sql record to update
+     */
+    public static void updateToTable(Connection connection, String tableName, Persistable persistable, int id, String currentSchema) {
+        upsertToTable(connection, tableName, persistable, id, currentSchema);
```

Comment:
The write calls to `SqlStorage` updates the [resource id](https://github.com/dimagi/formplayer/blob/master/src/main/java/org/commcare/formplayer/sandbox/SqlStorage.java#L125).
So if we write to `sqlite` first and then write to `postgres`, the [resource id would already be set](https://github.com/dimagi/formplayer/blob/master/src/main/java/org/commcare/formplayer/sandbox/SqlStorage.java#L119) and the `postgres` call to update the resource would fail because the resource was not  present in `postgres`. 

This workaround will use `insert on conflict` idea of `postgres`.

Reply:
I see Shivam made changes to this, is this now resolved? Also, if we fail in inserting something to postgres, that means that subsequent reads and potentially writes would fail?